Migrate keystore to master

pereyra-m opened the pull request:
|Related issue|
|---|
|Closes #26971|

<!--
This template reflects sections that must be included in new Pull requests.
Contributions from the community are really appreciated. If this is the case, please add the
"contribution" to properly track the Pull Request.

Please fill the table above. Feel free to extend it at your convenience.
-->

## Description

This PR migrates the `wazuh-keystore` to the `master` branch and adapts it to the project structure. The QA/UT tests and required utils were also migrated.

The OpenSSL v3.1.0 was added to vcpkg because it was missing and it was the closest version to the one used at branch v4.10.0.

The `IndexerConnector` class was updated to read the keystore and the plain text credentials were removed.

The SPECS and installer scripts were modified to include the new binary, but the RPM package can't be built due to previous errors.
The folders and binaries permissions were preserved but due to https://github.com/facebook/rocksdb/issues/11503 the internal RocksDB files will have 644 instead of 640.

Until this PR is merged https://github.com/wazuh/wazuh/issues/26908, the `root:wazuh` permissions will be used instead of  `wazuh:wazuh`.

## Tests

I've installed a package built from this branch and configured the keystore credentials.
A test event is properly indexed

![2024-11-27_20-44](https://github.com/user-attachments/assets/3d4688de-c6d2-41e7-91c2-012743067790)

pereyra-m commented on .github/workflows/engine_coverage.yml:
```diff
@@ -189,11 +189,11 @@ jobs:
 
             # Generate the coverage report
-            lcov  --no-external --capture --initial --directory ${module_build_path} --base-directory ${module_path} --exclude "${module_path}/test/*" --output-file ${coverage_base_file}
+            lcov  --no-external --capture --initial --directory ${module_build_path} --base-directory ${module_path} --exclude "${module_path}/test/*" --output-file ${coverage_base_file} --ignore-errors unused,mismatch
```
I've added this flags to ignore some errors:
- unused: the coverage reports that the keystore test folder exclusion is unused
- mismatch: the GTest macros create a number line difference that causes confusion to `lcov`

pereyra-m replied:
I've removed this.
Thank you @juliancnn

MiguelazoDS commented on src/engine/source/indexerconnector/src/indexerConnector.cpp:
```diff
@@ -118,16 +119,18 @@ static void initConfiguration(SecureCommunication& secureCommunication, const In
-    password = config.password;
 
-    if (config.username.empty())
+    const auto& kesytorePath = config.keystorePath.empty() ? KEYSTORE_PATH : config.keystorePath;
```
Typo multiple places: "kesy..."

pereyra-m replied:
Fixed!

MiguelazoDS commented on src/engine/source/keystore/src/keyStore.cpp:
```diff
@@ -0,0 +1,50 @@
+}
+
+/**
+ * Get the key value in the specified column family.
```
I think this no longer apply

MiguelazoDS commented on src/engine/source/base/include/base/utils/keyValueFile.hpp:
```diff
@@ -0,0 +1,224 @@
+            std::ofstream file(m_filePath);
+            if (!file.is_open())
+            {
+                throw std::runtime_error("Error creating key-value file due to: " + std::string(strerror(errno)));
```
Is it expect to fail if the folder is not created?

MiguelazoDS commented:
# Testing 

## Test tool

- Test tool does not work if the folder is not created 
```console
root@jammy:/home/vagrant/wazuh-5.0.0/src/engine/build/source/keystore# echo "secret_pass" | ./wazuh-keystore -k password
Error creating key-value file due to: No such file or directory
```
Probably expected, I couldn't find the creation of that directory in the code but is used prior this implementation.
- Quick test to check defered decryption 
Testtool was modified to return the decrypted value using d value in vp option. 
![image](https://github.com/user-attachments/assets/36071f8e-28de-4207-bf51-b89bfd6e19f5)

pereyra-m commented:
## Update

I've tested the decryption with Python

<details><summary>Script</summary>
<p>

```python
from Crypto.Cipher import AES

KEY_SIZE = 32
IV_SIZE = 16
KEYSTORE_PATH = 'key-value.bin'
KEY_VALUE_SEPARATOR = b':'

f = open(KEYSTORE_PATH, 'rb')
data = f.read()
f.close()

# Each key-value pair is stored in the following format:
# key:value_size:(enc_key+iv+cypher_text)
def get_key_value_map_from_file(data):
    pos = 0
    map = {}
    while pos < len(data):
        separator_1 = data.find(KEY_VALUE_SEPARATOR, pos)
        separator_2 = data.find(KEY_VALUE_SEPARATOR, separator_1 + 1)

        value_size = int(data[separator_1 + 1:separator_2])
        line_size = (separator_2 - pos) + 1 + value_size

        map[data[pos:separator_1]] = data[separator_2 + 1: separator_2 + 1 + value_size]
        pos += line_size + 1

    return map


map = get_key_value_map_from_file(data)

for key, value in map.items():
    enc_key = value[:KEY_SIZE]
    iv = value[KEY_SIZE:KEY_SIZE + IV_SIZE]
    value = value[KEY_SIZE + IV_SIZE:]

    cipher = AES.new(enc_key, AES.MODE_CBC, iv)
    decrypted = cipher.decrypt(value)

    print(f"{key.decode('utf-8')}:{decrypted.decode('utf-8')}")


```

</p>
</details> 

Values insertion
```
(venv) ╭─root@055969f33f50 /workspaces/5.x/wazuh 
╰─# ./src/engine/build/source/keystore/tools/wazuh-keystore_tool -k key1 -v myValue -p ./key-value.bin    
2024-12-04 23:38:59.991 391206:391206 keyValueFile.hpp:203 at put(): debug: Key 'key1' added to file './key-value.bin'.
Key 'key1' updated.
(venv) ╭─root@055969f33f50 /workspaces/5.x/wazuh 
╰─# ./src/engine/build/source/keystore/tools/wazuh-keystore_tool -k key2 -v 'Hello World!' -p ./key-value.bin
2024-12-04 23:39:25.842 391609:391609 keyValueFile.hpp:203 at put(): debug: Key 'key2' added to file './key-value.bin'.
Key 'key2' updated.
(venv) ╭─root@055969f33f50 /workspaces/5.x/wazuh 
╰─# ./src/engine/build/source/keystore/tools/wazuh-keystore_tool -k username -v 'admin' -p ./key-value.bin       
2024-12-04 23:39:39.085 391705:391705 keyValueFile.hpp:203 at put(): debug: Key 'username' added to file './key-value.bin'.
Key 'username' updated.
(venv) ╭─root@055969f33f50 /workspaces/5.x/wazuh 
╰─# ./src/engine/build/source/keystore/tools/wazuh-keystore_tool -k password -v 'e3äËL»vÙ*9ao"Éo=:`üË¼W+.ó]Â+\$' -p ./key-value.bin
2024-12-04 23:41:53.862 392252:392252 keyValueFile.hpp:203 at put(): debug: Key 'password' added to file './key-value.bin'.
Key 'password' updated.
```

File content
```
key1:64:R��"��-�#WQ�F��~ K��u�� �v@䤼�LtF���	P�#5�~Pd^[{;/���
key2:64:;�34��:�Z�#��3����9r�3�äͨO�N����ӽ_�Tʊ����_:}eq�r�{�W�
password:96:y
-�A���`�Z�(��#�7�T�!(��:�mA��_8�LJ��vǱ�@0@��[+���N����y2��J��;���f�p�:)�<|��
username:64:.��Q��/y���Πhn���I".���I}�IFL!�-Au�#he��brʇ
b'���f$��
```

Results
```

(venv) ╭─root@055969f33f50 /workspaces/5.x/wazuh 
╰─# python read_key-value.py
key1:myValue
key2:Hello World!
password:e3äËL»vÙ*9ao"Éo=:`üË¼W+.ó]Â+\$
username:admin
```

sebasfalcone reviewed (dismissed):
## Conclusion
The testing performed concludes that this feature is working not without some room for improvement, which should be discussed to be implemented in a proper manner.

- As of today, the encryption is not a robust security measure. Anyone that knows the key length will be able to decrypt this
    - We should add the possibility of using certificates 
- The storage engine is really rudimentary. It reads all the file and overwrites it with the changed data (we could implement a more efficient and complex system)    
    
> [!NOTE]
> We weren't able to test using packages due to a known issue https://github.com/wazuh/wazuh/issues/27138

pereyra-m commented:
## Update

This is the new Python script to read the whole file encrypted keystore

<details><summary>Details</summary>
<p>

```python
from Crypto.Cipher import AES

KEY_SIZE = 32
IV_SIZE = 16
KEYSTORE_PATH = 'key-value.bin'
KEY_VALUE_SEPARATOR = ':'

f = open(KEYSTORE_PATH, 'rb')
data = f.read()
f.close()

# The keystore is in the following format:
# key:value_size:(enc_key+iv+cypher_text)
enc_key = data[:KEY_SIZE]
iv = data[KEY_SIZE:KEY_SIZE + IV_SIZE]
value = data[KEY_SIZE + IV_SIZE:]
cipher = AES.new(enc_key, AES.MODE_CBC, iv)

def unpad(s):
    return s[:-ord(s[len(s)-1:])]

decrypted = unpad(cipher.decrypt(value)).decode('utf-8')

def get_key_value_map_from_file(decrypted):
    lines = decrypted.split('\n')
    lines.pop()
    map = {}
    for line in lines:
        key_value = line.split(KEY_VALUE_SEPARATOR)
        key = key_value[0]
        value = key_value[1]
        map[key] = value

    return map

map = get_key_value_map_from_file(decrypted)

for key, value in map.items():
    print(f"{key}:{value}")

```

</p>
</details>

MiguelazoDS commented:
# Testing (12/09) 

Due to some requirements changes we repeat some testing over this implementation. 

## RPM packages 
```console
[root@rhel8 vagrant]# rpm -ivh wazuh-server_5.0.0-0_x86_64_619ea42.rpm
Verifying...                          ################################# [100%]
Preparing...                          ################################# [100%]
Updating / installing...
   1:wazuh-server-5.0.0-0             ################################# [100%]
```
Binary exists
```console
[root@rhel8 vagrant]# /usr/share/wazuh-server/bin/wazuh-keystore --help
Switch value: -k not found.

Usage: wazuh-keystore <option(s)>
Options:
        -h                      Show this help message
        -k KEY                  Specifies the key for the key-value pair.
        -v VALUE                Specifies the value associated with the key. Only use one value option at the time.
        -vp VALUE_PATH          Path to a file containing the value to read (single line). Only use one value option at the time.
        -p KEYSTORE_PATH        Specifies the path to the keystore. Optional, the default value will be used if not provided.
        NOTE: if both value parameters are empty, stdin will be read.

Example:
        ./wazuh-keystore -k username -v admin

        ./wazuh-keystore -k username -v admin -p /custom/path/to/keystore

        ./wazuh-keystore -k password -vp /path/to/file.txt

        ./wazuh-keystore -k password < /path/to/file.txt

        echo 'pass' | ./wazuh-keystore -k password

        cat /path/to/file.txt | ./wazuh-keystore -k password
```

## DEB packages 
```console
root@jammy:/home/vagrant# dpkg -i wazuh-server_5.0.0-0_amd64_619ea42.deb
Selecting previously unselected package wazuh-server.
(Reading database ... 174439 files and directories currently installed.)
Preparing to unpack wazuh-server_5.0.0-0_amd64_619ea42.deb ...
Unpacking wazuh-server (5.0.0-0) ...
Setting up wazuh-server (5.0.0-0) ...
Processing triggers for libc-bin (2.35-0ubuntu3.8) ...
```

Binary exists
```console
root@jammy:/home/vagrant# /usr/share/wazuh-server/bin/wazuh-keystore --help
Switch value: -k not found.

Usage: wazuh-keystore <option(s)>
Options:
        -h                      Show this help message
        -k KEY                  Specifies the key for the key-value pair.
        -v VALUE                Specifies the value associated with the key. Only use one value option at the time.
        -vp VALUE_PATH          Path to a file containing the value to read (single line). Only use one value option at the time.
        -p KEYSTORE_PATH        Specifies the path to the keystore. Optional, the default value will be used if not provided.
        NOTE: if both value parameters are empty, stdin will be read.

Example:
        ./wazuh-keystore -k username -v admin

        ./wazuh-keystore -k username -v admin -p /custom/path/to/keystore

        ./wazuh-keystore -k password -vp /path/to/file.txt

        ./wazuh-keystore -k password < /path/to/file.txt

        echo 'pass' | ./wazuh-keystore -k password

        cat /path/to/file.txt | ./wazuh-keystore -k password
```

MiguelazoDS commented on .github/workflows/engine_indexer-connector_test.yml:
```diff
@@ -39,12 +39,12 @@ jobs:
+
       - name: Build
-        run: cmake --build ${{env.ENGINE_DIR}}/build --target indexer_connector_tool -j$(nproc)
+        run: cmake --build ${{env.ENGINE_DIR}}/build --target indexer_connector_tool wazuh-keystore -j$(nproc)
```
I think we can revert these changes now that the indexer connector won't talk directly to the keystore tool right?

pereyra-m replied:
You are right, I'll remove the this.

MiguelazoDS commented on src/engine/source/indexerconnector/qa/test_data/test_add_bulk_indexer_connector/config.json:
```diff
@@ -1,5 +1,6 @@
   "name": "wazuh-states-vulnerabilities",
   "enabled": "yes",
-  "hosts": ["http://localhost:9200"]
+  "hosts": ["http://localhost:9200"],
```
Do you still need this?

pereyra-m replied:
You are right, I'll remove the this.

MiguelazoDS commented on src/engine/source/indexerconnector/qa/test_data/test_initialize_indexer_connector/config.json:
```diff
@@ -1,5 +1,6 @@
   "name": "wazuh-states-vulnerabilities",
   "enabled": "yes",
-  "hosts": ["http://localhost:9200"]
+  "hosts": ["http://localhost:9200"],
+  "keystorePath" : "./keystore"
```
Do you still need this?

pereyra-m replied:
You are right, I'll remove the this.

MiguelazoDS commented on src/engine/source/indexerconnector/include/indexerConnector/indexerConnector.hpp:
```diff
@@ -38,7 +38,7 @@ struct IndexerConnectorOptions
     std::string username;           ///< The username to authenticate with OpenSearch.
     std::string password;           ///< The password to authenticate with OpenSearch.
-
+    std::string keystorePath;       ///< The path to the keystore.
```
Does the indexer connector need this?

pereyra-m replied:
You are right, I'll remove this.

MiguelazoDS commented on src/engine/source/conf/src/conf.cpp:
```diff
@@ -60,7 +60,8 @@ Conf::Conf(std::shared_ptr<IApiLoader> apiLoader)
     addUnit<int>(key::INDEXER_THREADS, "WAZUH_INDEXER_THREADS", 1);
     addUnit<std::string>(key::INDEXER_DB_PATH, "WAZUH_INDEXER_DB_PATH", "/var/lib/wazuh-server/indexer-connector/");
-
+    addUnit<std::string>(
+        key::INDEXER_KEYSTORE_PATH, "WAZUH_INDEXER_KEYSTORE_PATH", "/etc/wazuh-server/client.keystore");
```
Does the IC need this?

pereyra-m replied:
You are right, I'll remove this.

MiguelazoDS commented on src/engine/source/indexerconnector/qa/test_efficacy_log.py:
Do we really now need the changes in the indexer connector efficacy tests?

pereyra-m replied:
You are right, I'll remove this.

sebasfalcone commented:
Great job @pereyra-m @MiguelazoDS!

## Definitions
### Invalid keystore file
@MiguelazoDS Tested cases where the keystore contained invalid information, which the tool won't be able to decrypt therefore no data will be saved

The failure is expected in this scenario, because the alternative would be to overwrite the whole file, but the user may have entered an invalid path and lost important data

pereyra-m commented on src/engine/source/base/include/base/utils/homedirHelper.hpp:
I'll move this to fs.

pereyra-m commented:
## Update

Now the tool informs about the success or failure of the operation with more detail.

### No previous file

```bash
╭─root@055969f33f50 /workspaces/5.x/wazuh/src/engine/build/source/keystore 
╰─# ls client.keystore                                                                                                                      
ls: cannot access 'client.keystore': No such file or directory
╭─root@055969f33f50 /workspaces/5.x/wazuh/src/engine/build/source/keystore 
╰─# ./wazuh-keystore -k username -v admin -p ./client.keystore                                                                              
Key store file updated successfully.
```

### File with pre-existing data

```bash
╭─root@055969f33f50 /workspaces/5.x/wazuh/src/engine/build/source/keystore 
╰─# echo "data" > client.keystore 
╭─root@055969f33f50 /workspaces/5.x/wazuh/src/engine/build/source/keystore
╰─# ./wazuh-keystore -k username -v admin -p ./client.keystore
Unable to update key store file, no changes made. Reason: Invalid key store file encryption.
```

### File with not valid encryption 

```bash
╭─root@055969f33f50 /workspaces/5.x/wazuh/src/engine/build/source/keystore ‹enhancement/26971-migrate-keystore-tool●› 
╰─# cat client.keystore                                                                                                                   130 ↵
^[�-/�dr0^[u��6�a^[����\�^T�E��invalid^[�-/�dr0^[u��6�a^[����\�^T�E��
╭─root@055969f33f50 /workspaces/5.x/wazuh/src/engine/build/source/keystore ‹enhancement/26971-migrate-keystore-tool●› 
╰─# ./wazuh-keystore -k username -v admin -p ./client.keystore
Unable to update key store file, no changes made. Reason: Error finalizing decryption
```

sebasfalcone commented on src/engine/source/keystore/src/keyStore.cpp:
```diff
@@ -0,0 +1,117 @@
+
+constexpr auto KS_VALUE_SEPARATOR {':'}; // Default separator for key-value pairs.
+
+/**
+ * @brief Verifies if the file exists and creates it with the right permissions.
+ *
+ * @param filePath Path to the keystore.
+ */
```
Documentation goes in the hpp
```suggestion
```

pereyra-m replied:
Fixed!

sebasfalcone commented on src/engine/source/keystore/src/keyStore.cpp:
```diff
@@ -0,0 +1,117 @@
+    }
+}
+
+/**
+ * @brief Reads the content from the file and decrypts it.
+ *
+ * @param filePath File path.
+ * @return base::utils::KeyValue A key-value object.
+ */
```
Documentation goes in the hpp
```suggestion
```

pereyra-m replied:
Fixed!

sebasfalcone commented on src/engine/source/keystore/src/keyStore.cpp:
```diff
@@ -0,0 +1,117 @@
+    return base::utils::KeyValue(decryptedKeystoreStr, KS_VALUE_SEPARATOR);
+}
+
+/**
+ * @brief Insert or update a key-value pair in the keystore.
+ *
+ * @param key The key to be inserted or updated.
+ * @param value The value to be inserted or updated.
+ * @param keyStorePath The path to the key store file.
+ */
```
Documentation goes in the hpp
```suggestion
```

pereyra-m replied:
Fixed!

sebasfalcone commented on src/engine/source/keystore/src/keyStore.cpp:
```diff
@@ -0,0 +1,117 @@
+    outFile.close();
+}
+
+/**
+ * Get the a value from the keystore.
+ *
+ * @param key The key to be retrieved.
+ * @param value The returned value. Won't be modified if the key is not found.
+ * @param keyStorePath The path to the key store file.
+ */
```
Documentation goes in the hpp
```suggestion
```

pereyra-m replied:
Fixed!

sebasfalcone commented on src/engine/source/keystore/include/keyStore.hpp:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @param key The key to be inserted or updated.
+     * @param value The corresponding value to be returned.
+     * @param keyStorePath The path to the database file.
```
```suggestion
     * @param keyStorePath The path to the database file.
     * @return true If the key is found.
     * @return false If the key is not found.     
```

pereyra-m commented on src/engine/source/keystore/include/keyStore.hpp:
```diff
@@ -0,0 +1,62 @@
+    /**
+     * Get the value from the keystore.
+     *
+     * @param key The key to be inserted or updated.
```
Fixed!

sebasfalcone reviewed (dismissed):
- Code 🟢 
- Testing 🟢 

We verified that the tools behave as expected in a variety of scenarios:
- https://github.com/wazuh/wazuh/pull/27066#issuecomment-2528656168

The CLI and interfaces are comprehensive and easy to maintain

Dwordcito commented on src/engine/source/base/include/base/utils/keyValue.hpp:
```diff
@@ -0,0 +1,140 @@
+     *
+     * @return std::string The buffer with the key-value pairs.
+     */
+    std::string dumpMap()
```
This method could be const.
```suggestion
std::string dumpMap() const
````

pereyra-m replied:
Fixed

Dwordcito commented on src/engine/source/base/include/base/utils/keyValue.hpp:
```diff
@@ -0,0 +1,140 @@
+    const char m_separator;
+    const char m_newLine;
+    std::map<std::string, std::string> m_keyValueMap;
+
```
```suggestion
std::map<std::string, std::string, std::less<>> m_keyValueMap;
```

pereyra-m replied:
Fixed

Dwordcito commented on src/engine/source/base/include/base/utils/keyValue.hpp:
```diff
@@ -0,0 +1,140 @@
+     * @param value The value to be inserted or updated.
+     */
+    void put(const std::string& key, const std::string& value)
+    {
```
```suggestion
void put(const std::string& key, std::string_view value)
```

Dwordcito commented on src/engine/source/base/include/base/utils/keyValue.hpp:
```diff
@@ -0,0 +1,140 @@
+#include <filesystem>
+#include <fstream>
+#include <map>
+#include <sstream>
```
unused
```suggestion
```

pereyra-m replied:
Fixed

Dwordcito commented on src/engine/source/base/include/base/utils/keyValue.hpp:
```diff
@@ -0,0 +1,140 @@
+#ifndef _KEY_VALUE_HPP
+#define _KEY_VALUE_HPP
+
+#include <algorithm>
+#include <filesystem>
+#include <fstream>
```
unused

```suggestion

```

pereyra-m replied:
Fixed.

Dwordcito commented on src/engine/source/base/include/base/utils/keyValue.hpp:
```diff
@@ -0,0 +1,140 @@
+
+public:
+    KeyValue(const std::string& keyValueBuffer = "",
+             const char separator = KEY_VALUE_SEPARATOR,
```
```suggestion
explicit KeyValue(const std::string& keyValueBuffer = "",
```

pereyra-m replied:
Fixed

Dwordcito commented on src/engine/source/keystore/src/keyStore.cpp:
```diff
@@ -0,0 +1,92 @@
+constexpr auto KS_VALUE_SEPARATOR {':'}; // Default separator for key-value pairs.
+
+void Keystore::fileCreate(const std::string& filePath)
+{
```
```suggestion
void Keystore::fileCreate(const std::filesystem::path& filePath)
```

pereyra-m replied:
Fixed

Dwordcito commented on src/engine/source/keystore/include/keyStore.hpp:
```diff
@@ -0,0 +1,65 @@
+     * @param filePath Path to the keystore.
+     */
+    static void fileCreate(const std::string& filePath);
+
```
```suggestion
static void fileCreate(const std::filesystem::path& filePath);
```

pereyra-m replied:
Fixed

Dwordcito commented on src/engine/source/keystore/src/main.cpp:
```diff
@@ -0,0 +1,109 @@
+
+#include <filesystem>
+#include <fstream>
+#include <functional>
```
unused
```suggestion
```

pereyra-m replied:
Removed

Dwordcito commented on src/engine/source/keystore/src/argsParser.hpp:
```diff
@@ -0,0 +1,146 @@
+     *
+     * @return std::string Kesytore path.
+     */
+    std::string getKeystorePath() { return m_keystorePath; }
```
```suggestion
std::string getKeystorePath() const { return m_keystorePath; }
```

pereyra-m replied:
Fixed

Dwordcito commented on src/engine/vcpkg.json:
```diff
@@ -126,6 +127,10 @@
+    },
+    {
+      "name": "openssl",
+      "version": "3.1.0#0"
```
3.1.0 not is the latest version, we can use the 3.4.0#0 if we dont have some limitation to change the minor version, or we can use the 3.1.4#1 if we need maintain the minor version.

pereyra-m replied:
I selected the version closest to the one used in 4.x because of the OpenSSL helpers that I moved.

I'll try with 3.4.0#0

Dwordcito commented on src/engine/source/base/include/base/utils/keyValue.hpp:
```diff
@@ -0,0 +1,140 @@
+        if (key.find(m_separator) != std::string::npos || key.find(m_newLine) != std::string::npos
+            || value.find(m_separator) != std::string::npos || value.find(m_newLine) != std::string::npos)
+        {
+            throw std::runtime_error("Invalid key-value pair, can't contain separator or newline.");
```
what is the separator character? specify this in the throw message.

sebasfalcone reviewed (dismissed):
# Testing

## Insert key
```console
./wazuh-keystore -p ./client.key -k key                                                                                                                                  value1 
Key store file updated successfully.

tools/wazuh-keystore_tool -p ./client.key -k key 
2024-12-12 13:56:07.455 47718:47718 keyValue.hpp:78 at get(): debug: Key 'key' found.
key:value1
```

## Insert other key
```console
./wazuh-keystore -p ./client.key -k key2 -v value2                                                                                                                     Key store file updated successfully.

tools/wazuh-keystore_tool -p ./client.key -k key2                                                                                                                        
2024-12-12 13:56:05.858 47673:47673 keyValue.hpp:78 at get(): debug: Key 'key2' found.
key2:value2

tools/wazuh-keystore_tool -p ./client.key -k key 
2024-12-12 13:56:07.455 47718:47718 keyValue.hpp:78 at get(): debug: Key 'key' found.
key:value1
```


## Overwrite values
```console
tools/wazuh-keystore_tool -p ./client.key -k key 
2024-12-12 13:56:07.455 47718:47718 keyValue.hpp:78 at get(): debug: Key 'key' found.
key:newvalue

tools/wazuh-keystore_tool -p ./client.key -k key2 
2024-12-12 14:00:13.662 49458:49458 keyValue.hpp:78 at get(): debug: Key 'key2' found.
key2:newvalue
```

## Don't overwrite non keystore files
```console
./wazuh-keystore -p ./nonkeystore -k key2 -v value2                                                                                                                    
Unable to update key store file, no changes made. Reason: Invalid key store file encryption.
```

Dwordcito commented:
@pereyra-m is OOO, I made some changes in the Keystore::filecreate class, because we call the std::ofstream constructor with the filename and not with the filepath.

sebasfalcone reviewed (approved):
# Testing
## Insert values
```console
./wazuh-keystore -k stdin
stdin_value
Key store file updated successfully.

echo "echo_value" | ./wazuh-keystore -k echo      
Key store file updated successfully.

./wazuh-keystore -k file -vp value.txt 
Key store file updated successfully.

./wazuh-keystore -k arg -v arg_value  
Key store file updated successfully.
```

## Read values
```console
python3 read.py
arg:arg_value
echo:echo_value
file:file_value
stdin:stdin_value
```

## Overwrite values
```console
./wazuh-keystore -k arg -v arg_value_new
Key store file updated successfully.

python3 read.py
arg:arg_value_new
echo:echo_value
file:file_value
stdin:stdin_value
```
